### PR TITLE
Add fish shell completion

### PR DESCRIPTION
**Pull request by saschagrunert**
Hey there, I added the fish shell completion to podman. I also thought about adding an auto generation (cobra supports it for zsh, bash and (soon) for fish) subcommand, that the files don't have to be manually maintained. What do you think?

### CODE REVIEW

**Comment by vrothberg** [comment]
Thanks a lot, @saschagrunert!

>  I also thought about adding an auto generation (cobra supports it for zsh, bash and (soon) for fish) subcommand, that the files don't have to be manually maintained. What do you think?

I see one issue with auto-generating those from cobra, namely that it doesn't work for argument completions but only for commands, sub-commands, flags, and options. But we also need to complete image names, containers names, pods, etc.; any kind of completion that requires semantic knowledge.

I think auto-generated files are nice as a starting point, but I don't think we can avoid manual maintenance afterward.

File: Makefile [hunk 1]
```diff
@@ -28,6 +28,7 @@ OCI_RUNTIME ?= ""
 
 BASHINSTALLDIR=${PREFIX}/share/bash-completion/completions
 ZSHINSTALLDIR=${PREFIX}/share/zsh/site-functions
+FISHINSTALLDIR=${PREFIX}/share/fish/completion
```
Comment (vrothberg) [inline Makefile]: s/completion/completions/ ... at least on Fedora. Is it a typo or does openSUSE have a different path?

Comment (saschagrunert) [reply Makefile]: Your're right, it should be `completions`.

File: completions/fish/podman.fish [hunk 2]
```diff
@@ -0,0 +1,3201 @@
+
+function __fish_podman_no_subcommand --description 'Test if podman has yet to be given the subcommand'
+    for i in (commandline -opc)
+        if contains -- $i attach build commit container cp create diff events exec export generate generate-completion healthcheck history image images import info inspect kill load login logout logs mount pause play pod port ps pull push restart rm rmi run save search start stats stop system tag top umount unpause varlink version volume wait
```
Comment (TomSweeneyRedHat) [inline completions/fish/podman.fish]: maybe I've missed something, but is 'generate-completion' valid?

Comment (TomSweeneyRedHat) [reply completions/fish/podman.fish]: Oh, I think I see now, it's for the shell completions itself, not necessarily a podman command.

**Comment by TomSweeneyRedHat** [comment]
The changes LGTM, but I'm very concerned about manually maintaining podman.fish over time.  I can just see it getting out of step too quickly.   I'd feel more comfortable with it if podman.fish could be created reliably in a make process based on the Cobra settings.  Definitely want @mheon, @rhatdan and @baude to chime in on this one.

File: completions/fish/podman.fish (hunk 2 above)
Comment (saschagrunert) [reply completions/fish/podman.fish]: Yeah, sorry I forgot to remove it after my modification of podman :+1:

**Comment by saschagrunert** [comment]
We could wait for this one https://github.com/spf13/cobra/pull/754 and then add necessary fish functions for the completion of, for example, image names. Then we would have an auto-generated and manually maintained part of the completion.

**Comment by edsantiago** [comment]
@saschagrunert have you looked at the zsh completion code? That's almost entirely self-discovering, using `podman --help', with a few extensions for determining image/container/volume/registry names. Perhaps something like that could be done with fish as well? As others have mentioned, hardcoding every option is not maintainable.

And, incidentally, you will need to update `contrib/spec/podman.spec.in` to get CI to pass.

**Comment by saschagrunert** [comment]
> @saschagrunert have you looked at the zsh completion code? That's almost entirely self-discovering, using `podman --help', with a few extensions for determining image/container/volume/registry names. Perhaps something like that could be done with fish as well? As others have mentioned, hardcoding every option is not maintainable.

Yeah I also thought about an on-the-fly parser, but in the end I see pros and cons for both kinds of solutions. Let me elaborate how such a parser can be implemented and then come back with my findings.

/hold

**Comment by saschagrunert** [comment]
@edsantiago I played around with some scripted (sed-based) approach in fish shell and must say it is damn slow (some seconds) compared to a generated file. I really would like to recommend in generating the completion grammatically and injecting necessary static functions as constant strings. The generation of the shell completion could then be done via some make target. WDYT?

I guess since the related PR in fish shell is not merged this can't be done here, too... :worried:

**Comment by edsantiago** [comment]
Autogenerating a completion file at build time was actually my first approach; I abandoned it for several reasons including the difficulty of validating the generated results. I have to assume that your PR is mostly autogenerated; if you can come up with a reliable script that will run at build time, using only tools that are available on build systems, I'd be much more inclined to accept it enthusiastically. Thank you again for your efforts on this.

**Comment by saschagrunert** [comment]
I will close this for now and will follup up on this topic in the future.

**Pull request closed without merging**